### PR TITLE
W-11855052: Database Connector to connect to SQL Anywhere via the sajdbc4 driver

### PR DESCRIPTION
**Pull request by alopezbizon**

### CODE REVIEW

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/MuleApplicationClassLoader.java [hunk 1]
```diff
@@ -24,17 +28,34 @@
...
 public class MuleApplicationClassLoader extends MuleDeployableArtifactClassLoader {
 
   static {
     registerAsParallelCapable();
   }
 
   private final NativeLibraryFinder nativeLibraryFinder;
+  private final LazyValue<Class<?>> librariesLoaded = new LazyValue<>();
+  private final LazyValue<Class<?>> dynamicLibraryLoader = new LazyValue<>(this::getDynamicLibraryLoader);
+  public static final String METHOD_NAME = "loadLibrary";
```
Comment (IvanAndresFritzler) [inline modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/MuleApplicationClassLoader.java:47]: Move to line 43

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/MuleApplicationClassLoader.java [hunk 2]
```diff
@@ -45,6 +66,11 @@ public MuleApplicationClassLoader(String artifactId, ArtifactDescriptor artifact
 
   @Override
   protected String findLibrary(String name) {
+    if (!this.librariesLoaded.isComputed()) {
+      this.librariesLoaded.get();
+      loadNativeLibraries(name);
```
Comment (IvanAndresFritzler) [inline modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/MuleApplicationClassLoader.java:71]: We can name this method as "loadNativeLibraryDependencies"

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/MuleApplicationClassLoader.java [hunk 3]
```diff
@@ -130,4 +156,26 @@ private static List<ClassLoader> resolveContextArtifactPluginClassLoadersForCurr
 
     return classLoaders;
   }
+
+  private void loadNativeLibraries(String nativeLibraryNameExcluded) {
```
Comment (IvanAndresFritzler) [inline modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/MuleApplicationClassLoader.java:160]: Add a //TODO with the tickets that can improve this.

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/MuleApplicationClassLoader.java [hunk 4]
```diff
@@ -130,4 +156,26 @@ private static List<ClassLoader> resolveContextArtifactPluginClassLoadersForCurr
...
+
+    List<String> nativeLibraries = nativeLibraryFinder.getNativeLibraries();
+    nativeLibraries.remove(nativeLibraryNameExcluded);
+    Collections.reverse(nativeLibraries);
+
+    for (String nativeLibrary : nativeLibraries) {
+      try {
+        method.invoke(loader.getDeclaredConstructor().newInstance(), nativeLibrary);
+      } catch (IllegalAccessException | InvocationTargetException | InstantiationException | NoSuchMethodException e) {
+        throw new MuleRuntimeException(e);
```
Comment (IvanAndresFritzler) [inline modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/MuleApplicationClassLoader.java:177]: Improve this with an exception message.

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/ArtifactCopyNativeLibraryFinder.java [hunk 5]
```diff
@@ -13,15 +13,20 @@
...
 import java.io.IOException;
 import java.net.URISyntaxException;
 import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
 import java.util.Optional;
 
 import org.apache.commons.io.FilenameUtils;
+import org.mule.runtime.api.exception.MuleRuntimeException;
```
Comment (IvanAndresFritzler) [inline modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/ArtifactCopyNativeLibraryFinder.java:29]: Imports order.

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeLibraryFinder.java [hunk 6]
```diff
@@ -20,4 +22,11 @@ public interface NativeLibraryFinder {
    * @return library path to use for the given name. Can be null is no library was found.
    */
   String findLibrary(String name, String libraryPath);
+
+  /**
+   * Provides the shared native libraries
+   *
+   * @return the shared native libraries. Can be an empty list if no native library is found.
+   */
+  List<String> getNativeLibraries();
```
Comment (IvanAndresFritzler) [inline modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeLibraryFinder.java:31]: Add a TODO that will be deleted after the same tasks that we added to the ApplicationClassloader.

Comment (IvanAndresFritzler) [inline modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeLibraryFinder.java:31]: Rename to findLibraryNames

**Review by IvanAndresFritzler — CHANGES_REQUESTED** [review]

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeExtension.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+ */
+
+package org.mule.runtime.module.artifact.activation.internal.nativelib;
+
+public enum NativeExtension {
+
+  DLL, DYLIB, JNILIB, SO;
+
+  @Override
+  public String toString() {
```
Comment (IvanAndresFritzler) [inline modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeExtension.java:15]: Rename to value

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeExtension.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+
+package org.mule.runtime.module.artifact.activation.internal.nativelib;
+
+public enum NativeExtension {
```
Comment (IvanAndresFritzler) [inline modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeExtension.java:10]: Rename to NativeLibraryFileExtension

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeLibraryFinder.java [hunk 9]
```diff
@@ -20,4 +22,12 @@ public interface NativeLibraryFinder {
...
    */
   String findLibrary(String name, String libraryPath);
+
+  /**
+   * Provides the native library names
+   *
+   * @return the native library names. Can be an empty list if no native library name is found.
+   */
+  List<String> findLibraryNames();
+  // TODO: W-12786373 Delete after do the same task added in the MuleApplicationClassloader.
```
Comment (IvanAndresFritzler) [inline modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeLibraryFinder.java:32]: // TODO: Delete after completing W-12786373

**Review by IvanAndresFritzler — CHANGES_REQUESTED** [review]

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/NativeLibraryLoaderMuleDeployableArtifactClassLoader.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+  protected void loadNativeLibraryDependencies(String nativeLibraryNameExcluded) {
+    // TODO: W-12786373 Implement loading native libraries with dependencies.
+    if (!dynamicLibraryLoader.isComputed()) {
+      Class<?> loader = dynamicLibraryLoader.get();
+      Method method;
+      try {
+        method = loader.getMethod(METHOD_NAME, String.class);
+      } catch (NoSuchMethodException e) {
+        throw new MuleRuntimeException(
+                                       createStaticMessage("Can not obtain the %s method to load the native library dependencies.",
```
[the remark this answers is not in the document]
Comment (alopezbizon) [reply modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/NativeLibraryLoaderMuleDeployableArtifactClassLoader.java:80]: The MuleRuntimeException only has 3 public constructors: `MuleRuntimeException(I18nMessage message)`, `MuleRuntimeException(I18nMessage message, Throwable cause)` and `MuleRuntimeException(Throwable cause)`. The constructor `MuleRuntimeException(String message, Throwable cause)` is private, so I can't use `String.format()`.

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/NativeLibraryLoaderMuleDeployableArtifactClassLoader.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+  protected NativeLibraryLoaderMuleDeployableArtifactClassLoader(String artifactId, ArtifactDescriptor artifactDescriptor,
+                                                                 ClassLoader parentCl,
+                                                                 NativeLibraryFinder nativeLibraryFinder, List<URL> urls,
+                                                                 ClassLoaderLookupPolicy lookupPolicy) {
+    super(artifactId, artifactDescriptor, urls.toArray(new URL[0]), parentCl, lookupPolicy);
+    this.nativeLibraryFinder = nativeLibraryFinder;
+    this.supportNativeLibraryDependencies = isFeatureEnabled(SUPPORT_NATIVE_LIBRARY_DEPENDENCIES, artifactDescriptor);
+  }
+
+  protected void loadNativeLibraryDependencies(String nativeLibraryNameExcluded) {
```
[the remark this answers is not in the document]
Comment (alopezbizon) [reply modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/NativeLibraryLoaderMuleDeployableArtifactClassLoader.java]: Done.

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeExtension.java (hunk 7 above)
Comment (alopezbizon) [reply modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeExtension.java]: Done.

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeLibraryFinder.java (hunk 9 above)
Comment (alopezbizon) [reply modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeLibraryFinder.java]: Done.

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeExtension.java (hunk 8 above)
Comment (alopezbizon) [reply modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/nativelib/NativeExtension.java]: Done.

**Comment by alopezbizon** [comment]
--validate

**Comment by alopezbizon** [comment]
--validate

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/NativeLibraryLoaderMuleDeployableArtifactClassLoader.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (alopezbizon) [reply modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/NativeLibraryLoaderMuleDeployableArtifactClassLoader.java]: Done.

File: modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/NativeLibraryLoaderMuleDeployableArtifactClassLoader.java [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+      List<String> nativeLibraries = nativeLibraryFinder.findLibraryNames();
+      nativeLibraries.remove(nativeLibraryNameExcluded);
+      Collections.reverse(nativeLibraries);
+
+      for (String nativeLibrary : nativeLibraries) {
+        try {
+          method.invoke(loader.getDeclaredConstructor().newInstance(), nativeLibrary);
+        } catch (IllegalAccessException | InvocationTargetException | InstantiationException | NoSuchMethodException e) {
+          throw new MuleRuntimeException(
```
[the remark this answers is not in the document]
Comment (alopezbizon) [reply modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/NativeLibraryLoaderMuleDeployableArtifactClassLoader.java]: Done.

**Review by IvanAndresFritzler — APPROVED** [review]

**Pull request merged**